### DISCUSSION TITLE
Git workflow

### DISCUSSION THREAD

**Discussion by almirsarajcic** [3Stones-io/omedis#311] [General]
### Latest version
When you're working on a new task, check out the `main` branch first, then `git pull`.
After that, create a new branch off of that. That way, you're basing your branch on the latest version of the codebase preventing unnecessary issues.

### Branch naming
When naming a branch, it's helpful to start with a number of the GitHub issues, e.g. `235-add-registration-via-invites`. That enables you to easily check out that branch by typing `git checkout 235-` and pressing the tab (at least in the Zsh shell with oh-my-zsh installed). You can still utilize comprehensive names, which can help when you're looking at the list of branches, but also when searching for a branch locally with `grep` or something.
<img width="1254" alt="Screenshot 2024-11-09 at 18 28 19" src="https://github.com/user-attachments/assets/4745f68d-4ef3-4cbe-9c47-c8b4250c404a">

### Collaborating on a branch
Be careful with the Push/Sync/Publish button in your IDE. It's not wired to the GitHub so it doesn't get updates automatically until you do some operations with it (like fetching), so you might think you're creating the branch for the first time, but it might already exist in the repo on GitHub.
<img width="482" alt="Screenshot 2024-11-09 at 18 18 45" src="https://github.com/user-attachments/assets/727d95e3-83bf-4c32-82e8-0ed92b0a0eb9">
That can happen in some situations that I can't remember right now, but trust me, they exist.

### Merging
When you want to update your feature branch to include the changes in the `main` branch, check out the `main` branch, then `git pull`, then check out your branch again and do `git merge main`. If you do `git pull origin main` directly, you might get into weird issues (@kagure-nyakio felt this multiple times this week).

One of the issues you might get into with that is that your `main` branch is never up to date with the upstream, so whenever you start new feature branches based on `main`, you're already behind (see "Latest version").

When you get merge conflicts and start to resolve them, Merge Editor is not your friend. When it's a simple merge conflict that's basically just choosing one or the other version, you can use it, but for anything else, click on the file in the list in your IDE and use the git diff shown to resolve the issue. Since it's a plain text, Cursor Tab can help you most of the time. When you're ready to commit the merge, don't forget to stage the unstaged files from the end of the list. We missed this so many times, and undoing merge commits is at least an order of magnitude harder than regular commits.

**Comment by wintermeyer** [comment]
I found https://gitbutler.com to be a very helpful git GUI.

**Reply by almirsarajcic** [reply]
Niiice! I thought it was a basic GUI (like GitHub Desktop), but it seems it's much more.

**Reply by wintermeyer** [reply]
> Niiice! I thought it was a basic GUI (like GitHub Desktop), but it seems it's much more.

For me Gitbutler is a real time saver.
